README.md
![license](https://img.shields.io/github/license/loliot/lot-SBC)
![version](https://img.shields.io/github/v/tag/loliot/lot-SBC?sort=semver)
![language](https://img.shields.io/github/languages/top/loliot/lot-SBC)

# lot-API

## Adding New board

```bash
git clone https://github.com/loliot/lot-SBC.git
```

Change **lot-SBC** to **lot-\<your board\>** in `debian/*` and Doxyfile.
Make the board name a combination of lowercase letters and hyphens.

Rewrite README.md.

```bash
git remote rename origin lot-SBC &&\
git remote add origin https://github.com/loliot/lot-<your board>.git &&\
git add . &&\
git commit -m "init" &&\
git push -f -u origin master
```

## Pin mapping

Pin mapping is based on physical pinout of board.

### Example

In case using an 8-bit processor and using PA3, PA4, PC0, PC2, PD1, and PD4.

Set port A, C, and D numbers to 0, 16, and 32, respectively.

Set PA3 and PA4 to 3(0+3) and 4(0+4).  
Set PC0 and PC2 to 16(16+0) and 18(16+2).  
Set PD1 and PD4 to 33(32+1) and 36(32+4).

```bash
|     | LOT |  PHY  | LOT |     |
| PA3 |   3 | 1 | 2 |   4 | PA4 |
| PD4 |  36 | 3 | 4 |  16 | PC0 |
| PC2 |  18 | 5 | 6 |  33 | PD1 |
```

## Structure

```bash
lot-<your borad>
├── lot-API
│   ├── II2c.h
│   ├── ISpi.h
│   ├── IUart.h
│   ├── Log.h
│   ├── lotdef.h
│   ├── lot_gpio.h
│   ├── lot_time.h
│   └── ...
├── SBC.h
├── lot.h
├── Gpio.h
├── I2c.h
├── Spi.h
├── Uart.h
└── ...
```

## Editing Makefile

If adding a.cpp source to lot-\<your board\>, add the source name to CXXSRCS in Makefile as shown below.

```make
CSRCS = \

SXXSRCS = a.cpp \

INCS_DIR = .\

```

## Code formatting

```bash
make clang
```


debian/changelog
lot-SBC (0.12.3) unstable; urgency=medium

  * API: Rename enum identifier to mixed case without '_'

 -- Hyeonki Hong <hhk7734@gmail.com>  Thu, 30 Jan 2020 20:50:38 +0900

lot-SBC (0.12.2) unstable; urgency=medium

  * API: Add virtual destructor
  * API: Add override keyword
  * API: Move functions using cs_pin into ISpi.hpp
  * API: Add using-declaration to introduce members of IGpio

 -- Hyeonki Hong <hhk7734@gmail.com>  Wed, 29 Jan 2020 14:54:28 +0900

lot-SBC (0.12.1) unstable; urgency=medium

  * API: Use -O2 instead of -O1 and remove -g

 -- Hyeonki Hong <hhk7734@gmail.com>  Thu, 23 Jan 2020 12:16:38 +0900

lot-SBC (0.12.0) unstable; urgency=medium

  * API: Correct parameter order
  * API: Move init() to constructor

 -- Hyeonki Hong <hhk7734@gmail.com>  Sat, 18 Jan 2020 01:42:46 +0900

lot-SBC (0.11.4) unstable; urgency=medium

  * API: Refactor Gpio class
  * API: Update docs
  * API: Change type of 'start_millis/micros' to 'std::atomic'

 -- Hyeonki Hong <hhk7734@gmail.com>  Fri, 17 Jan 2020 16:58:03 +0900

lot-SBC (0.11.3) unstable; urgency=medium

  * API: Move typedef into loddef.h

 -- Hyeonki Hong <hhk7734@gmail.com>  Wed, 08 Jan 2020 10:11:52 +0900

lot-SBC (0.11.2) unstable; urgency=medium

  * API: Fix bug where mode() does not set m_mode

 -- Hyeonki Hong <hhk7734@gmail.com>  Mon, 06 Jan 2020 17:12:55 +0900

lot-SBC (0.11.1) unstable; urgency=medium

  * API: Add PWM pin_mode_t
  * API: Use Gpio.digital instead of gpio::digital
  * API: Remove gpio()

 -- Hyeonki Hong <hhk7734@gmail.com>  Sun, 05 Jan 2020 23:19:52 +0900

lot-SBC (0.11.0) unstable; urgency=medium

  * API: Refactor type definition
  * API: Update license
  * API: Remove LOT mode variables
  * API: Remove set/get_pin_speed
  * API: Remove set_, get_, _write, _read
  * API: Move GPIO functions into gpio namespace
  * API: Implement Gpio class
  * API: Add xx_stoi, xx_itos in lotdef.h

 -- Hyeonki Hong <hhk7734@gmail.com>  Fri, 03 Jan 2020 23:14:16 +0900

lot-SBC (0.10.0) unstable; urgency=medium

  * API: Fix debian install error
  * API: Implement get_lot_pin_available(pin)
  * API: Use throw to handle errors instead of exit()
  * API: Implement unsupported_error class

 -- Hyeonki Hong <hhk7734@gmail.com>  Thu, 02 Jan 2020 10:59:54 +0900

lot-SBC (0.9.2) unstable; urgency=medium

  * API: Refactor Log

 -- Hyeonki Hong <hhk7734@gmail.com>  Tue, 31 Dec 2019 16:38:51 +0900

lot-SBC (0.9.1) unstable; urgency=medium

  * API: Add pkg-config file

 -- Hyeonki Hong <hhk7734@gmail.com>  Tue, 31 Dec 2019 13:53:26 +0900

lot-SBC (0.9.0) unstable; urgency=medium

  * API: Add INFO log level
  * API: Move *_linux.cpp into linux directory
  * API: Prevent reopen when already open

 -- Hyeonki Hong <hhk7734@gmail.com>  Sun, 29 Dec 2019 13:29:54 +0900

lot-SBC (0.8.3) unstable; urgency=medium

  * API: Set default argument
  * API: Add checking if pin is available
    Pin mux was not considered completely.

 -- Hyeonki Hong <hhk7734@gmail.com>  Tue, 24 Dec 2019 00:36:57 +0900

lot-SBC (0.8.2) unstable; urgency=medium

  * API: Remove infinite loop in Log::error

 -- Hyeonki Hong <hhk7734@gmail.com>  Sun, 17 Nov 2019 15:27:14 +0900

lot-SBC (0.8.1) unstable; urgency=medium

  * API: Add delay_us_2() (delay using hard loop)

 -- Hyeonki Hong <hhk7734@gmail.com>  Sat, 09 Nov 2019 14:25:06 +0000

lot-SBC (0.8.0) unstable; urgency=medium

  * API: Add lot_time_init()
  * API: Implement lot_time for Linux

 -- Hyeonki Hong <hhk7734@gmail.com>  Sat, 09 Nov 2019 12:58:13 +0000

lot-SBC (0.7.0) unstable; urgency=medium

  * API: Implement Log

 -- Hyeonki Hong <hhk7734@gmail.com>  Fri, 08 Nov 2019 22:25:23 +0900

lot-SBC (0.6.0) unstable; urgency=medium

  * API: Implement SPI source for Linux
  * API: Add source template

 -- Hyeonki Hong <hhk7734@gmail.com>  Thu, 31 Oct 2019 14:35:36 +0000

lot-SBC (0.5.0) unstable; urgency=medium

  * API: Implement I2C source for Linux

 -- Hyeonki Hong <hhk7734@gmail.com>  Tue, 22 Oct 2019 23:49:38 +0900

lot-SBC (0.4.0) unstable; urgency=medium

  * API: Implement ios, ostream class
  * API: Implement UART source for Linux

 -- Hyeonki Hong <hhk7734@gmail.com>  Sat, 12 Oct 2019 15:28:54 +0900

lot-SBC (0.3.1) unstable; urgency=medium

  * API: Test debian package

 -- Hyeonki Hong <hhk7734@gmail.com>  Wed, 06 Oct 2019 09:03:00 +0900

lot-SBC (0.3.0) unstable; urgency=medium

  * API: Test Makefile
  * API: Change source tree
  * API: Add I2C, UART, SPI header

 -- Hyeonki Hong <hhk7734@gmail.com>  Wed, 03 Oct 2019 13:50:00 +0900

lot-SBC (0.2.0) unstable; urgency=medium

  * API: Add GPIO header, UART interface, Doxyfile

 -- Hyeonki Hong <hhk7734@gmail.com>  Thu, 26 Sep 2019 14:48:00 +0900

lot-SBC (0.1.0) unstable; urgency=medium

  * API: Initial release.
  * API: Add I2C, SPI, time interface

 -- Hyeonki Hong <hhk7734@gmail.com>  Thu, 22 Sep 2019 22:53:00 +0900


lot_SBC.h
/*
 * MIT License
 * Copyright (c) 2019-2020 Hyeonki Hong <hhk7734@gmail.com>
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#include "lot-API/lotdef.h"

#define MAX_PHY_PIN_COUNT 40
#define MAX_LOT_PIN_COUNT 36

namespace lot
{
const int phy_to_lot[MAX_PHY_PIN_COUNT + 1] = {
    // clang-format off
     UNUSED,
     UNUSED, UNUSED,   // 1   | 2
     UNUSED, UNUSED,   // 3   | 4
     UNUSED, UNUSED,   // 5   | 6
     UNUSED, UNUSED,   // 7   | 8
     UNUSED, UNUSED,   // 9   | 10
     UNUSED, UNUSED,   // 11  | 12
     UNUSED, UNUSED,   // 13  | 14
     UNUSED, UNUSED,   // 15  | 16
     UNUSED, UNUSED,   // 17  | 18
     UNUSED, UNUSED,   // 19  | 20
     UNUSED, UNUSED,   // 21  | 22
     UNUSED, UNUSED,   // 23  | 24
     UNUSED, UNUSED,   // 25  | 26
     UNUSED, UNUSED,   // 27  | 28
     UNUSED, UNUSED,   // 29  | 30
     UNUSED, UNUSED,   // 31  | 32
     UNUSED, UNUSED,   // 33  | 34
     UNUSED, UNUSED,   // 35  | 36
     UNUSED, UNUSED,   // 37  | 38
     UNUSED, UNUSED    // 39  | 40
    // clang-format on
};

const int lot_to_phy[MAX_LOT_PIN_COUNT] = {
    // clang-format off
     UNUSED, UNUSED,   // |
     UNUSED, UNUSED,   // |
     UNUSED, UNUSED,   // |
     UNUSED, UNUSED,   // |
     UNUSED, UNUSED,   // |
     UNUSED, UNUSED,   // |
     UNUSED, UNUSED,   // |
     UNUSED, UNUSED,   // |
     UNUSED, UNUSED,   // |
     UNUSED, UNUSED,   // |
     UNUSED, UNUSED,   // |
     UNUSED, UNUSED,   // |
     UNUSED, UNUSED,   // |
     UNUSED, UNUSED,   // |
     UNUSED, UNUSED,   // |
     UNUSED, UNUSED,   // |
     UNUSED, UNUSED,   // |
     UNUSED, UNUSED    // |
    // clang-format on
};

const bool is_available_lot[MAX_LOT_PIN_COUNT] = {
    // clang-format off
     false, false,   // |
     false, false,   // |
     false, false,   // |
     false, false,   // |
     false, false,   // |
     false, false,   // |
     false, false,   // |
     false, false,   // |
     false, false,   // |
     false, false,   // |
     false, false,   // |
     false, false,   // |
     false, false,   // |
     false, false,   // |
     false, false,   // |
     false, false,   // |
     false, false,   // |
     false, false    // |
    // clang-format on
};
}    // namespace lot